### ISSUE TITLE
Machine Functions

### ISSUE THREAD

**Issue by moonheart08** [moonheart08/SERAIS#2]
Old idea I've been sitting on.
`CALLMFN imm15 | Call machine function.`
`CHKMFN imm15 | Check if machine function is present, writing the C flag.` 
Essentially, optional CISC-y-ness for domain or application specific purposes. Programs can safely fallback to a software implementation if a MFN is not present, and presence checks *should* be cheap.
An example potential usecase is hardware assisted memcpy/memmove, where the check has minimal runtime overhead and implementations that choose to provide such machine functions likely can do it faster than software could.

**Comment by chorman0773** [comment]
> An example potential usecase is hardware assisted memcpy/memmove, where the check has minimal runtime overhead and implementations that choose to provide such machine functions likely can do it faster than software could.

The overhead could even be removed, and could be implemented similar to cpuid flags, in that the compiler could be told what operations it can rely upon, and it just doesn't care if they don't at runtime(=UB), then it's set by machine-specific flags, and/or tested using `-march=native`.

**Comment by moonheart08** [comment]
> > An example potential usecase is hardware assisted memcpy/memmove, where the check has minimal runtime overhead and implementations that choose to provide such machine functions likely can do it faster than software could.
> 
> The overhead could even be removed, and could be implemented similar to cpuid flags, in that the compiler could be told what operations it can rely upon, and it just doesn't care if they don't at runtime(=UB), then it's set by machine-specific flags, and/or tested using `-march=native`.

Yea, of course, if the compiler knows what it's targeting supports it, the overhead is nil. Implementations can likely relegate recognizing CHKMFN and CALLMFN validity to a "complex" decoder unit, as well. It's unlikely an implementation would need to decode multiple consecutive calls, so translating to internal ucode in decode would work, as the called function index is an immediate. CHKMFN could easily just become a uop that sets or unsets the carry flag, which can be merged with a consecutive branch.

**Comment by chorman0773** [comment]
One question I have, what would happen if you called a machine function that doesn't exist? Fault?

**Comment by moonheart08** [comment]
> One question I have, what would happen if you called a machine function that doesn't exist? Fault?

Yes, exactly. I forgot to mention this in the issue. Unallocated MFNs are supposed to fault on call. A section of the ID space will be reserved to spec, while the other portion will be reserved to vendors, and may require more complex detection mechanism due to the limited space a 16-bit int provides.

**Comment by Evrey** [comment]
So, what's the difference between a »machine function« and just having a BIOS? I.e. a piece of system software that exposes a bunch of function entry points to call. If context switching is what you fear, all you need to do is make this BIOS run in the current privilege level. Call it BIOS, user-land interrupt vector, system library,…

Essentially this sounds like what you want is hardware-support for mapping an OS-provided DLL into every address space. You don't even need µcoding, as that µcode would just end up being your readily decoded normal RISC code. So again no difference to calling a function.

**Comment by moonheart08** [comment]
@Evrey Using a simple BIOS ROM is in fact a valid implementation, and that's intended. However I intend for mfns to be used for extremely common operations like memcpy, where a CPU may be able to provide a faster, hardware level implementation of the operation. (say, moving entire cache lines at once, something user code cannot do)

This is also why they're entirely optional, an implementation that has no benefits from providing these can simply not include any mfns, hardwire CALLMFN to always fault, and hardwire CHKMFN to always indicate no such function exists.
They are not intended to be a special, OS provided "do OS thing" instruction, and won't provide such functions. They are intended to accelerate common, low level operations that *might* have faster hardware implementations.

**Comment by moonheart08** [comment]
Another way to think of it is an extremely granular way of extending the ISA.

**Comment by Evrey** [comment]
Well, makes sense.

An alternative would be having all the building blocks instructions one would need exposed, and offering easy to understand and use interrupts for emulating unimplemented instructions in software.

Say you had a special cache line register extension. Gives you, say, two cache line registers which you can use for load/store, and maybe very trivial masking. Then you also add a `memcpy` ISA extension that requires the cache line extension. And then you just compile your code normally.

Userland executes `memcpy`. If it exists in hardware, it's run in hardware, job done. Maybe using µcode, who knows. But if it doesn't exist in hardware or µcode, then a software interrupt is triggered requesting `memcpy` emulation. Now, if you have the cache line extension, your software emulation can do the super fast cache line juggling, as if it was µcode. But if that also doesn't exist, software emulation can emulate regular cache line instructions with multiple 64-bit load/stores, and specialise the `memcpy` for these smaller loads as well.

It's highly forwards- and backwards-compatible. Just expose whatever you think you'd need for µcode to the software-visible ISA and see where things would go. Having cache line transfer instructions would allow for super agressive inlining of quite big data transfers.

**Comment by moonheart08** [comment]
The main reason I'm opting for this method is it's completely agnostic to the hardware. We could completely reinvent the way cache works and mfn memcpy would still work exactly as one would expect.
The main thing that makes me set it up so that interrupts are never used is simply because, well, interrupts require interrupting the instruction stream. You can jump around an unimplemented mfn for quite a bit cheaper than triggering an interrupt.
